### PR TITLE
Restrict invalid expressions starting with a simple type

### LINKED ISSUE #26253: Bad sad moment when passing byte[] to a names argument
**Description:**

Definition : 
```java
public type TargetType typedesc<Response|string|xml|json|byte[]|CustomRecordType| CustomRecordType[]>;

public remote function post(string path, TargetType targetType = Response) returns Response|PayloadType|ClientError {
}
```
Usecase : 
```ballerina
var res = clientEP->post("/backend/getByteArray", byte[]);
```

internal log exception : 
```
[2020-10-07 12:50:34,342] SEVERE {b7a.log.crash} - org.wso2.ballerinalang.compiler.tree.expressions.BLangTypedescExpr cannot be cast to org.wso2.ballerinalang.compiler.tree.expressions.BLangVariableReference 
java.lang.ClassCastException: org.wso2.ballerinalang.compiler.tree.expressions.BLangTypedescExpr cannot be cast to org.wso2.ballerinalang.compiler.tree.expressions.BLangVariableReference
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:2587)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangIndexBasedAccess.accept(BLangIndexBasedAccess.java:56)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:361)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:340)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:4142)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangNamedArgsExpression.accept(BLangNamedArgsExpression.java:65)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:361)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:340)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkTypeParamExpr(TypeChecker.java:5425)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkTypeParamExpr(TypeChecker.java:5418)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInvocationArgs(TypeChecker.java:5239)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInvocationParam(TypeChecker.java:5170)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInvocationParamAndReturnType(TypeChecker.java:5118)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkActionInvocation(TypeChecker.java:5088)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:2692)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation$BLangActionInvocation.accept(BLangInvocation.java:218)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:361)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:340)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1756)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangAssignment.accept(BLangAssignment.java:81)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2759)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:2755)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:374)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:58)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:346)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2759)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:2751)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeClassDefinition(SemanticAnalyzer.java:434)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:423)
	at org.wso2.ballerinalang.compiler.tree.BLangClassDefinition.accept(BLangClassDefinition.java:106)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2759)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:2751)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:269)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.lambda$visit$2(SemanticAnalyzer.java:279)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:279)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:158)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:243)
	at org.wso2.ballerinalang.compiler.CompilerDriver.typeCheck(CompilerDriver.java:329)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:258)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:250)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:142)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:176)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:147)
	at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:69)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.TestCommand.execute(TestCommand.java:434)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.tool.Main.main(Main.java:57)

```

**Affected Versions:**
slp4

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
$Subject

Fixes #26253

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -3828,6 +3828,30 @@ private STNode parseExpressionRhsInternal(OperatorPrecedence currentPrecedenceLe
                     isInConditionalExpr);
         }
 
+        if (isSimpleTypeNode(lhsExpr.kind)) {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:3831]: I think parsing is correct in this scenario. `byte[]` is same as `Person[]`. Both are index-access expressions.
i.e:
```ballerina
type Person record {|
|}

public function main() {
    var x = Person[]; // ----(1)
    var y = byte[]; // ----(2)
}
```
Both (1) and (2) are identical, in terms of syntax and semantics.

If we consider either of (1) 0r (2), compiler should give two errors:
- Parser would give an error for missing index.
- Semantic analyzer would give an error saying `invalid operation: type 'typedesc<int>' does not support indexing`

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:3831]: I think this needs to be fixed at semantic analyzer

**Review by SupunS — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -2584,6 +2584,12 @@ private boolean isXmlAccess(BLangFieldBasedAccess fieldAccessExpr) {
 
     public void visit(BLangIndexBasedAccess indexBasedAccessExpr) {
         // First analyze the variable reference expression.
+        if (indexBasedAccessExpr.expr instanceof  BLangTypedescExpr) {
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:2587]: Shall we use `NodeKind` to check the type of the expression?

**Review by SupunS — CHANGES_REQUESTED** [review]

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/CustomFunctionTest.java [hunk 3]
```diff
@@ -43,4 +43,12 @@ public void testDuplicateParameters() {
         BAssertUtil.validateError(compile, 0, "redeclared symbol 'param'", 1, 35);
     }
 
+    @Test(description = "Test defining ballerina function with extra parameters with missing comma")
+    public void testExtraParameters() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/CustomFunctionTest.java:47]: I'm not sure if the name/description is accurate?

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/extra-parameters.bal [hunk 4]
```diff
@@ -0,0 +1,20 @@
+//  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/functions/extra-parameters.bal:1]: We use hyphens in module names and underscores in bal file names, shall we fix this?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/CustomFunctionTest.java [hunk 5]
```diff
@@ -43,4 +43,12 @@ public void testDuplicateParameters() {
...
+        CompileResult compile = BCompileUtil.compile("test-src/functions/extra-parameters.bal");
+        Assert.assertEquals(compile.getErrorCount(), 2);
+        BAssertUtil.validateError(compile, 0, "invalid operation: type 'byte' does not support indexing", 20, 13);
+        BAssertUtil.validateError(compile, 1, "missing key expr in member access expr", 20, 18);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/CustomFunctionTest.java:51]: Shall we change this to use an index and validate the error count?

```suggestion
        int index = 0;
        BAssertUtil.validateError(compile, index++, "invalid operation: type 'byte' does not support indexing", 20, 13);
        BAssertUtil.validateError(compile, index++, "missing key expr in member access expr", 20, 18);
        Assert.assertEquals(compile.getErrorCount(), index);
```

That way we can validate that we only get the expected errors.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/CustomFunctionTest.java [hunk 6]
```diff
@@ -43,4 +43,12 @@ public void testDuplicateParameters() {
         BAssertUtil.validateError(compile, 0, "redeclared symbol 'param'", 1, 35);
     }
 
+    @Test(description = "Test defining ballerina function with extra parameters with missing comma")
+    public void testExtraParameters() {
+        CompileResult compile = BCompileUtil.compile("test-src/functions/extra-parameters.bal");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/CustomFunctionTest.java:48]: How about something like

```suggestion
        CompileResult compileResult = BCompileUtil.compile("test-src/functions/extra-parameters.bal");
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 2 above)
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:2587]: Shall we also extract `indexBasedAccessExpr.expr` out to a variable? Seems to be used over and over again.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/CustomFunctionTest.java (hunk 3 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/CustomFunctionTest.java:47]: The issue is when an array-type-descriptor is used as an expression right? Not an issue with parameters?

For example, even the following fails

```ballerina
function foo(any a) returns int {
    return 1;
}

typedesc t = byte[];

int a = foo(t);
```

Shall we rename and move this test to `TypedescTests` maybe?

Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/CustomFunctionTest.java:47]: +1 will move

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/bytetype/BByteArrayValueNegativeTest.java [hunk 7]
```diff
@@ -67,5 +66,8 @@ public void testBlobValueNegative() {
         BAssertUtil.validateError(result, index++, "missing plus token", 16, 23);
         BAssertUtil.validateError(result, index++, "invalid base64 content in byte array literal", 17, 23);
         BAssertUtil.validateError(result, index++, "invalid base64 content in byte array literal", 18, 23);
+        BAssertUtil.validateError(result, index++, "invalid operation: type 'byte' does not support indexing", 19, 18);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/bytetype/BByteArrayValueNegativeTest.java:69]: Let's move this test to org.ballerinalang.test.types.typedesc.TypedescTests as discussed. Can you also add another test similar to

```ballerina
public function main() {
    typedesc t = int[]|string[];
}
```

Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/bytetype/BByteArrayValueNegativeTest.java:69]: sure

**Review by MaryamZi — APPROVED** [review]

**Review by SupunS — APPROVED** [review]

**Pull request merged**